### ISSUE TITLE
Next release 2.0.0

### ISSUE THREAD

**Issue by NickNaso** [nodejs/node-addon-api#570]
Hi everyone,
it's my intention create a new release. The last release is 1.7.1 and was deployed the 24rd July. Now we have some new APIs and important fixes to release. Here I report what we need to create this new release:

- [x] **Add Env() to AsyncContext** - Merge PR https://github.com/nodejs/node-addon-api/pull/568
- [x] **scopes: make failure of closing scopes fatal** - Merge PR https://github.com/nodejs/node-addon-api/pull/566
- [x]  **scope: disallow copying scopes to prevent double close of scopes** - Merge PR https://github.com/nodejs/node-addon-api/pull/565
- [x] **tsfn: Add wrappers for Ref and Unref** - Merge PR https://github.com/nodejs/node-addon-api/pull/561 (We need to merge PR https://github.com/nodejs/node-addon-api/pull/546 before)
- [x] **Update the CMake.js doc to current best practices** - Merge PR https://github.com/nodejs/node-addon-api/pull/558
- [x] **doc: fix return type of ArrayBuffer::Data** - Merge PR https://github.com/nodejs/node-addon-api/pull/552
- [x] **tsfn: Implement copy constructor** - Merge PR https://github.com/nodejs/node-addon-api/pull/546 (We need to merge this before the https://github.com/nodejs/node-addon-api/pull/561)
- [x] **Update Package JSON formatting** Update the formatting so that we use the same format
as it will be in once published.
- [x] **doc: Added AsyncProgressWorker to readme** - Merge PR https://github.com/nodejs/node-addon-api/pull/588
- [x] **Fix example code in doc/class_property_descriptor** - Merge PR https://github.com/nodejs/node-addon-api/pull/587

Nice to have:
- [x] **AsyncProgressWorker** - End work on PR https://github.com/nodejs/node-addon-api/pull/529 and merge it.
- [x] **implement Object::AddFinalizer** -  End work on PR https://github.com/nodejs/node-addon-api/pull/551 and merge it.
- [x] **NAPI_EXPERIMENTAL mixed with various Node.js Major Version** - End work on PR https://github.com/nodejs/node-addon-api/pull/545 and merge it.
- [x] fix `ThreadSafeFunction::[Non]BlockingCall` with pre-constructed napi_tsfn: Review and merge PR https://github.com/nodejs/node-addon-api/pull/580
- [x] add error handling to `ThreadSafeFunction::GetContext`: merge PR https://github.com/nodejs/node-addon-api/pull/583 . See https://github.com/nodejs/node-addon-api/pull/583#issuecomment-551008873 regarding semver-major release discussion

To discuss: 

- ~~-  **build: remove Node.js 6 from and add Node.js 13 to travis** - Merge PR https://github.com/nodejs/node-addon-api/pull/573~~ deferred until 3.0.0

If anyone have some other things that needs to be done please add them in the list above.

EDIT 2019-10-29T19:10:16.385Z KevinEady: added #580 to "Nice to Have"
EDIT 2019-11-15T21:50:07.887Z KevinEady: added #583 to "Nice to Have"

**Comment by KevinEady** [comment]
Hi @NickNaso ,

I am currently working on a PR to address https://github.com/nodejs/node-addon-api/issues/556#issuecomment-544634867 like we discussed yesterday. I am waiting on #546 copy ctor before pushing it though. I think this is a good addition for 2.0.0 as well. I will add it to the list once I create a PR.

**Comment by legendecas** [comment]
https://github.com/nodejs/node-addon-api/pull/547 introduced a breaking change on the usage of the definition of NAPI_EXPERIMENTAL, it would be nice to land such a change on the semver major version.

https://github.com/nodejs/node-addon-api/pull/545 is a related one.

**Comment by NickNaso** [comment]
> #545 introduced a breaking change on the usage of the definition of NAPI_EXPERIMENTAL, it would be nice to land such a change on the next semver major version.

@legendecas I added #545  to the list.

**Comment by KevinEady** [comment]
If this is a major version release, can we get rid of https://github.com/nodejs/node-addon-api/blob/master/napi-inl.deprecated.h ...?

**Comment by gabrielschulhof** [comment]
@KevinEady good point!

**Comment by NickNaso** [comment]
@KevinEady @gabrielschulhof could we remove the file without consequences? Is it right?

**Comment by gabrielschulhof** [comment]
@NickNaso there were also some conditionally defined sections that needed removing.

https://github.com/nodejs/node-addon-api/pull/572

**Comment by NickNaso** [comment]
> @NickNaso there were also some conditionally defined sections that needed removing.
> 
> #572

Added to the list with #573 .

**Comment by mhdawson** [comment]
Not sure we want to do all: remove deprecated stuff but keep infra - Merge PR #572.  While technically we can because its a SemVer major, we decided to do a major only because of changes we believe will have not effect on existing users.  My preference would to avoid trully breaking changes.

**Comment by gabrielschulhof** [comment]
I'm OK with deferring the removal of the deprecated stuff. I've struck out the PR from the checklist.

**Comment by legendecas** [comment]
Just a thought: maybe we could utilize labels and GitHub milestone/project to track the releasing process. :P

**Comment by arielweinberger** [comment]
Can't wait for @legendecas' pull request bringing AsyncProgressWorker to node-addon-api!

**Comment by arielweinberger** [comment]
Can't wait for @legendecas' pull request bringing AsyncProgressWorker to node-addon-api! Thank you everybody!

**Comment by extmchristensen** [comment]
Really excited about this new release (in particular with "tsfn: Implement copy constructor" which is a key feature that really makes a diff). I just hope the orginal scope is not increased too much so it gets delayed.

**Comment by mhdawson** [comment]
@NickNaso I think we are pretty much ready to make the new release.

**Comment by NickNaso** [comment]
@mhdawson I will wait until tomorrow just in case someone want add other things in the release.

**Comment by NickNaso** [comment]
Hi everyone I'm starting to create and publish a new release ...

**Comment by NickNaso** [comment]
Hi @gabrielschulhof @mhdawson, 
I prepared all to publish the new release but we have problems on Node.js version 11 https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api-new/1067/. I think that I can procede to publish on npm. Is it ok to you?

**Comment by gabrielschulhof** [comment]
@NickNaso it looks like the CI is having trouble running Node.js v11.

**Comment by mhdawson** [comment]
@NickNaso at this point 11 is long out of service.  The failures also look to me like issues with the configuration of the job (versus anything in the tests) that I probably only fixed up for 12 and later.

From my perspective those failures are 11 are not a concern.

**Comment by mhdawson** [comment]
The one on 13 was also likely a machine issue but started a new one for 13 to be sure.

**Comment by NickNaso** [comment]
Hi @mhdawson,
now we have failures for all Node.js version on AIX, but I think that is an infrastructure problem: 
![ci](https://user-images.githubusercontent.com/387506/69476657-dffbaa80-0ddc-11ea-9c1e-2e83eaf96a7f.png)
Could someone help o solve this problem?

**Comment by NickNaso** [comment]
The CI is green now and I will proceed to publish the new release on **npm**.

**Comment by mhdawson** [comment]
Those might still be machine issues, but the problem is that the other machines are off line so we'll need to get that figured out @sam-github

**Comment by NickNaso** [comment]
The CI now is green:
- **[Node.js v8](https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api-new/1171/)**  ✓
- **[Node.jsv10](https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api-new/1170/)** ✓
- **[Node.js v12](https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api-new/1172/)** ✓
- **[Node.js v13](https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api-new/1174/)** ✓
- **[Node.j v14 (master)](https://ci.nodejs.org/view/x%20-%20Abi%20stable%20module%20API/job/node-test-node-addon-api-new/1173/)** ✓
I'm going to complete the work to publish the release.

**Comment by NickNaso** [comment]
I'm closing the issue because the new release has been published.
